TaskFramework.md

```diff
@@ -0,0 +1,5 @@
+# Task Framework
+I want you to design a task framework. This framework will trigger 1 billion tasks everyday. A task can be defined as a code executable. We want our task framework to support task runners in all modern programming languages. This task framework also needs to have a topological ordering, in which we can trigger subtasks and the sub tasks can trigger subtasks. Some tasks may be dependent on specific things but that is up for the user of our framework to implement. 
+
+# Some constraints
+A task of a given level in the task tree needs to be executing all possible tasks at the level similanously as to insure speed. we also need to be able to view results or metadata on these tasks and their execution. Keep in mind this is a framework that will be integrated and used in an existing architecture so we want to prioritze as little footprint as possible without sacrificing performance.
```

Comment:
8. We can run tasks simultaneously, but is there any guarantee that they are idempotent? Is there any guarantee that they won't clash with the same data store? (e.g. what happens if task A wants to write X to a file, but task B wants to write Y to the same file. If task A gets the file first, and then task B overwrites the file, did task A succeed?) Should we do some sort of namespacing for each task to avoid this? If so, we'd sacrifice speed and memory, since we'd have to spin up an independent environment for each task.
9. What sort of metadata should we be able to see? Like progress/success/failure, logging? What else would be necessary?
10. How much hardware will be available to us? Since we want to run a large number of jobs in parallel, we'll probably need many machines, hopefully stateless (although tasks might not be). If so, what happens if a machine dies or partially fails (hard disk corruption might happen at a large enough scale -- we expect 3 faulty drives per 100 ssds/year, so if we need 100 machines to run tasks for a year, there's going to be some non-zero percent of failure.

Reply:
8. Assume that the datastores these tasks are dependent are configured in such a way that they are incredibly scalable and will not have any issues with writes from mulitple sources.
9. Success, Created, Pending, Failure thats up to you use some enum to represent the different task node states. 
10. Assume we have a large datacenter with an api for provisioning resources and we are only responsible for writing the software for managing how much of these resources we use.